src/useTowerGame.ts

```diff
@@ -1,103 +1,130 @@
 import { useFrame } from '@react-three/fiber';
 
-import { TOWER_BOUNDS, STARTING_SPEED, START_DIMENSIONS, BOX_HEIGHT } from './constants';
-import { DIRECTION, StackingBox } from './Types';
+import { TOWER_BOUNDS, STARTING_SPEED, BOX_HEIGHT } from './constants';
+import { Controls, DIRECTION, StackingBox } from './Types';
+import { useKeyboardControls } from '@react-three/drei';
+import { TowerActionType, useTowerReducer, initialBoxes } from './towerReducer';
+import { boxIsOverLappingNeg, boxIsOverLappingPos } from './utils';
 
 export const useTowerGame = () => {
-  const lastBox = useRef<Mesh>();
-  const movingBox = useRef<Mesh>();
-  const [movingBoxDimesions] = useState(START_DIMENSIONS);
-  const [direction, setDirection] = useState<DIRECTION>();
+    const lastBox = useRef<Mesh>();
+    const movingBox = useRef<Mesh>();
+    const [{ direction, movingBoxStartingPosition, boxes, movingBoxDimesions }, dispatch] =
+    useTowerReducer();
 
-  const boxes: Array<StackingBox> = [
-    {
-      position: [0, 0, 0],
-      args: [START_DIMENSIONS.width, BOX_HEIGHT, START_DIMENSIONS.length],
-      color: 'darkcyan',
-    },
-    {
-      position: [0, 1 * BOX_HEIGHT, 0],
-      args: [START_DIMENSIONS.width, BOX_HEIGHT, START_DIMENSIONS.length],
-      color: 'cyan',
-    },
-    {
-      position: [0, 2 * BOX_HEIGHT, 0],
-      args: [START_DIMENSIONS.width, BOX_HEIGHT, START_DIMENSIONS.length],
-      color: 'DarkGreen',
-    },
-  ];
+    const [wasPressed, setWasPressed] = useState(false);
+    const spacePressed = useKeyboardControls<Controls>((state) => state.hit);
+
+    const stackNewBox = useCallback(() => {
+        const lastBoxEdges = new Box3().setFromObject(lastBox.current);
+        const movingBoxEdges = new Box3().setFromObject(movingBox.current);
+        let newWidth = lastBoxEdges.max.x - lastBoxEdges.min.x;
+        let newLength = lastBoxEdges.max.z - lastBoxEdges.min.z;
+        let newStartX = lastBoxEdges.min.x + newWidth / 2;
+        let newStartZ = lastBoxEdges.min.z + newLength / 2;
+
+        if (boxIsOverLappingPos('x', lastBoxEdges, movingBoxEdges)) {
+            newWidth = lastBoxEdges.max.x - movingBoxEdges.min.x;
+            newStartX = movingBoxEdges.min.x + newWidth / 2;
+        } else if (boxIsOverLappingNeg('x', lastBoxEdges, movingBoxEdges)) {
+            newWidth = movingBoxEdges.max.x - lastBoxEdges.min.x;
+            newStartX = lastBoxEdges.min.x + newWidth / 2;
+        } else if (boxIsOverLappingPos('z', lastBoxEdges, movingBoxEdges)) {
+            newLength = lastBoxEdges.max.z - movingBoxEdges.min.z;
+            newStartZ = movingBoxEdges.min.z + newLength / 2;
+        } else if (boxIsOverLappingNeg('z', lastBoxEdges, movingBoxEdges)) {
+            newLength = movingBoxEdges.max.z - lastBoxEdges.min.z;
+            newStartZ = lastBoxEdges.min.z + newLength / 2;
+        } else {
+            dispatch({ type: 'END_GAME' });
+
+            return;
+        }
+
```

Comment:
Potentially a function that return the new position and args. Then return false if end of game.